### DISCUSSION TITLE
Issue with sort plugin and x-for

### DISCUSSION THREAD

**Discussion by nleter** [alpinejs/alpine#4157] [1. Help] [answered]
I have issue with x-for and sorting the elements.

I have an array that I want to sort using alpine x-sort plugin. Issue is that each element of the array has an x-model attribute to allow modification of that element.

Once I move element around, the array updates correctly but since the original element is moved in dom, x-model render the wrong position in the array.

I know that the issue is with using the index with x-model. 

Codepen for reference: https://codepen.io/nlet/pen/qBwJJzg

**Comment by ekwoka** [comment]
You need to use unique keys on the x-for.

**Reply by Arifursdev** [reply]
having unique keys for both x-for and x-sort doesn't work either. there is no example in the docs. do you have an example code that's working with x-for?

it updates the object in the correct order but broken visually.

**Reply by Arifursdev** [reply]
Both are Broken. or Is it really broken or am i doing something wrong here 🤔? @ekwoka

```html
<div x-data="{ 
    tasks: ['Task 1', 'Task 2', 'Task 3'],
    reorderTasks(item, newPosition) {
        let task = this.tasks[item];
        this.tasks.splice(item, 1);
        this.tasks.splice(newPosition, 0, task);
    }
}">
            <div x-sort="(item, newPosition) => reorderTasks(item, newPosition)">
                <template x-for="(task, index) in tasks" :key="index">
                    <div class="task" x-sort:item="index" x-sort:handle>
                        <p x-text="task"></p>
                    </div>
                </template>
            </div>
        </div>
```

Broken:
```html
        <br>
<div x-data="{ 
            tasks: [
                {id: 1, name: 'Task 1'}, 
                {id: 2, name: 'Task 2'}, 
                {id: 3, name: 'Task 3'}
            ],
            reorderTasks(item, newPosition) {
                let task = this.tasks[item];
                this.tasks.splice(item, 1);
                this.tasks.splice(newPosition, 0, task);
            }
        }">
            <div x-sort="(item, newPosition) => reorderTasks(item, newPosition)">
                <template x-for="(task, index) in tasks" :key="task.id">
                    <div class="task" x-sort:item="index" x-sort:handle>
                        <p x-text="task.name"></p>
                    </div>
                </template>
            </div>
        </div>
```

**Reply by marttheis** [reply]
Was anyone able to find a solution to this yet? I doubt that they would release a broken plugin like that. reordering columns in a alpineJS x-for table is the most obvious application beyond the more simple stuff.

**Reply by hugozap** [reply]
I think there should be a warning added to the AlpineJS documents because the `sort` plugin won't work with x-for.
The problem lies in the x-for directive having an internal variable that keeps track of the order of items `_x_prevKeys` .  Even if you update your array in the handler x-sort, the internal _x_prevKeys won't be updated in the `template` element and the order of items in the UI will differ.

You can solve the issue by re-setting that variable in the element that has the `x-for` (usually a template), it should contain the list of item "keys". So it should be usually done after applying the sorting changes to the array (when the sorting operation completes).

I think, the x-sort plugin should do that by default when it detects it's being used in combination with a x-for because it's the most common scenario.

**Reply by ekwoka** [reply]
:thinking: it does make a lot of sense

**Reply by Arifursdev** [reply]
@hugozap 
> You can solve the issue by re-setting that variable in the element that has the `x-for` (usually a template), it should contain the list of item "keys". So it should be usually done after applying the sorting changes to the array (when the sorting operation completes).

with plugin as it is, do you have an example code of how you would solve the issue with x-for?

> I think, the x-sort plugin should do that by default when it detects it's being used in combination with a x-for because it's the most common scenario.

truly!

Thanks!

**Reply by hugozap** [reply]
@Arifursdev Here's an small example using sortable.js directly. I tried to use the sort plugin but had issues with the scope in the handler function. The idea is the same though but i recommend just using sortable.js as the plugin seems broken to me.

https://codesandbox.io/p/sandbox/alpine-xfor-sort-fix-pj2z7n?file=%2Findex.html%3A1%2C1-65%2C1

(If you comment the line that sets _x_prevKeys and sort the items and then try to add a new one, the new position will be incorrect. This also happens with the plugin.

**Reply by Arifursdev** [reply]
@hugozap that makes sense! Thanks!

**Reply by babeuloula** [reply]
@hugozap Thanks you for the codesandbox! I had the same issue and now it's fixed 🙏🏼

**Comment by romanberdnikov** [comment]
The sort plugin also looks broken to me (just because I can't access this.$store directly). 

Another approach is to have a "sort" button or up/down buttons near every element.
Doesn't look pretty for user, but does it's job.

**Up/Down buttons:**
```
<div class="items-sort">
   <button x-show="$store.items.length > 1 && index > 0" @click="moveUp(index)" type="button">
   ^
   </button>
   <button x-show="$store.items.length > 1 && (index + 1) != $store.items.length" @click="moveDown(index)" type="button">
   v
   </button>
</div>
```
```
document.addEventListener("alpine:init", () => {
        init() {
            Array.prototype.move = function (from, to) {
                this.splice(to, 0, this.splice(from, 1)[0]);
            };
        },
        moveUp(index) {
            this.$store.items.move(index, index - 1);
        },
        moveDown(index) {
            this.$store.items.move(index, index + 1);
        },
});
```

**Sort button:**

```
<button @click="sortItems" type="button">
Sort by time
</button>
```
```
document.addEventListener("alpine:init", () => {
        sortItems() {
            this.$store.items.sort(function (a, b) {
                return a.time - b.time; /* time - any other key in array */
            });
        }
});
```

**Reply by ekwoka** [reply]
> The sort plugin also looks broken to me (just because I can't access this.$store directly).

Can you clarify what you mean by this?

It feels like that move won't work all the time, since it will be changing the indexes when it moves things.......but I guess moving one at a time it works fine

but I'd probably prefer

```diff
- Array.prototype.move = function (from, to) {
+ Array.prototype.moveOne = function (from, offset) {
-   this.splice(to, 0, this.splice(from, 1)[0]);
+  [this[from+offset],this[from] = [this[from],this[from+offset]];
};
```

**Reply by romanberdnikov** [reply]
I mean I can't access Alpine "this" inside x-sort="handle", it only receives "item, position".
I can change the data directly by calling Alpine.data(), but it doesn't affect DOM (afair) and also the problem with _x_prevKeys.

It's working fine:
https://codepen.io/romanberdnikov/full/KKLeopz

**Reply by Arifursdev** [reply]
@romanberdnikov as the example given by @hugozap

this works: 
https://codesandbox.io/p/sandbox/alpine-xfor-sort-fix-pj2z7n?file=%2Findex.html%3A1%2C1-65%2C1

**Reply by romanberdnikov** [reply]
> @romanberdnikov as the example given by @hugozap

There is no delete button

**Reply by Arifursdev** [reply]
> There is no delete button

```js
deleteItem(index) {
    this.$store.items.splice(index, 1);
}
```

**Reply by romanberdnikov** [reply]
You can't access Alpine from the sort plugin directly. So after sorting items, they actually won't change their order (only visually in DOM).

Delete functionality brings another problem with the _x_prevKeys. It won't be able to find the last stored key

**Reply by khanhicetea** [reply]
Thanks, updating _x_prevKeys of template element does the trick.

My try : https://jsbin.com/jijirigiya/edit?html,console,output

**Reply by kevariable** [reply]
@khanhicetea you save my time !, thank you so much. after struggle 2 days on this

**Comment by ekwoka** [comment]
This is solved with https://github.com/alpinejs/alpine/pull/4361

**Comment by dimitri-koenig** [comment]
Hi guys

I finally found a solution that works without issues so far. The trick is to give random :keys in the x-for template tag. I do this by incrementing a sortIteration variable every time the sort handler is called. That way a rerender is forced.

```html
<div x-data="columnSorting">
    <ul x-sort="handleSortOrder">
        <template x-for="column in selectedColumns" :key="sortIteration + column">
            <li x-sort:item="column">
                <span x-text="column"></span>
            </li>
        </template>
    </ul>
</div>
```

```js
Alpine.data('columnSorting', () => {
    let self;

    return {
        selectedColumns: ['item1', 'item2', 'item3'],

        sortIteration: 0,

        init() {
            self = this;
        },

        handleSortOrder(item, newIndex) {
            let currentIndex = self.selectedColumns.indexOf(item);

            self.selectedColumns.splice(currentIndex, 1);
            self.selectedColumns.splice(newIndex, 0, item);

            self.sortIteration++;
        },
    }
});
```

Since I don't get access to the alpine store data object using "this", I have to declare a self variable outside of the object. Hacky workaround, but works. Hope that helps anyone.

**Reply by IMythx** [reply]
you could access the `this` without the need for the `self` (or the way you did it) by binding the `handle` function to `$data` magic variable which is basically the `Alpine.data`
```html
  <ul x-sort="handleSortOrder.bind($data)">
        ...
  </ul>
```

**Comment by amadesclaire** [comment]
This works for me

```html
<ul x-sort="$store.myStore.handleSort">
  <template x-for="(item) in $store.myStore.items" :key="item.id + '-' + item.order">
    <li x-sort:item="{ item }">
    ...
    </li>
  </template>
</ul>
```
In handleSort() I make a POJO of the items array, manually update the item.order then overwrite the store's version. Working fine so far.

**Reply by brianfeister** [reply]
This fix worked for me

**Comment by bennadel** [comment]
Just ran into this as well. I'm setting the `:key="Math.floor( Math.random() * 999999 )"` to get around the issue. I know this is causing DOM elements to be created/destroyed , but not sure if there's another way around it.

**Reply by ebacala** [reply]
Hacky indeed but it works, I'm using this at the moment, thanks! 😄

**Comment by calebporzio** [comment, marked as answer]
this has been fixed in v3.15.9 and later. tyty

**Reply by Arifursdev** [reply]
Thank you!
